### PR TITLE
Add support for OpenCV 4.2

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**

### CODE REVIEW

File: CMake/fletch-tarballs.cmake
Comment (dstoup) [inline CMake/fletch-tarballs.cmake:305]: Maybe not the time to tackle this issue but do we really need to keep 2.4.13 around?

**Review by dstoup — APPROVED** [review]
Looks good to me. My comment about version 2 isn't really necessary to address at this time but I bring it up but adding new version is usually when I start dropping old versions as well. 

The only other thing that I might suggest is to temporarily set 4.2.0 as the default version to see if the branch actually works across platforms. If you want to keep the integrity of this branches history, you could pull a branch from this HEAD, update the default and make a PR. Once we see the results, you could just delete that temp PR.

**Comment by dstoup** [comment]
Jenkins test this please

**Comment by dstoup** [comment]
> Jenkins test this please

The Windows issues seems to be a download from sourceforge problem .. go figure.

File: CMake/fletch-tarballs.cmake
Comment (mwoehlke-kitware) [reply CMake/fletch-tarballs.cmake:305]: That is an excellent question which I do not consider myself qualified to answer :slightly_smiling_face:. Maybe a question for @mattdawkins and @mleotta?

**Comment by mwoehlke-kitware** [comment]
> The only other thing that I might suggest is to temporarily set 4.2.0 as the default version to see if the branch actually works across platforms.

Yeah, that sounds useful... expect a (separate, "testing only") PR soon.

Depending on your definition of "work", I know KWIVER *won't* compile against 4.x, which was sort of the motivation for this. (To wit: Fedora does not ship 3.x and it's annoying that KWIVER can't be built with Fedora's packaged version. I seem to have that locally patched, but pushing those patches is problematic with fletch not also being able to provide 4.x.)

**Comment by mleotta** [comment]
I'm in favor of updating to OpenCV 4 both here and in KWIVER.  If for no other reason than because OpenCV 3 does not support CUDA 11 (and I don't think it's worth patching).  So holding back OpenCV also holds back our CUDA version.  I think one reason we haven't moved to OpenCV 4 yet is because there is likely some non-trivial work involved to update the KWIVER arrow.

@mwoehlke-kitware does this branch mean that you are also working on a KWIVER branch to update KWIVER to use OpenCV 4.2?

**Comment by dstoup** [comment]
> > The only other thing that I might suggest is to temporarily set 4.2.0 as the default version to see if the branch actually works across platforms.
> 
> Yeah, that sounds useful... expect a (separate, "testing only") PR soon.
> 
> Depending on your definition of "work", I know KWIVER _won't_ compile against 4.x, which was sort of the motivation for this. (To wit: Fedora does not ship 3.x and it's annoying that KWIVER can't be built with Fedora's packaged version. I seem to have that locally patched, but pushing those patches is problematic with fletch not also being able to provide 4.x.)

Yeah, I agree that KWIVER should be better about accepting system packages. I planned to start a pattern for that and see if anyone objects. So for me, 'work' in this case really just means that it builds in Fletch without errors. I figured KWIVER (or vidtk for that matter) won't build against it yet.

Thanks!

**Comment by mwoehlke-kitware** [comment]
> I think one reason we haven't moved to OpenCV 4 yet is because there is likely some non-trivial work involved to update the KWIVER arrow. @mwoehlke-kitware does this branch mean that you are also working on a KWIVER branch to update KWIVER to use OpenCV 4.2?

Ahem: "I seem to have [KWIVER] locally patched"....

```
$ git diff --stat
 arrows/ocv/bounding_box.h                        | 12 ++++++------
 arrows/ocv/detect_features_AGAST.cxx             |  7 ++++---
 arrows/ocv/detect_features_FAST.cxx              |  4 ++--
 arrows/ocv/detect_heat_map.cxx                   |  2 +-
 arrows/ocv/detect_motion_3frame_differencing.cxx |  4 ++--
 arrows/ocv/draw_detected_object_set.cxx          |  2 +-
 arrows/ocv/estimate_fundamental_matrix.cxx       |  2 +-
 arrows/ocv/feature_detect_extract_ORB.cxx        |  7 ++++---
 arrows/ocv/hough_circle_detector.cxx             |  4 ++--
 arrows/ocv/image_container.cxx                   | 10 +++++-----
 arrows/ocv/tests/test_bounding_box.cxx           |  2 +-
 arrows/proj/geo_conv.cxx                         |  1 +
 12 files changed, 30 insertions(+), 27 deletions(-)
```

It builds and the tests pass. I can't promise that's "working", though.

This was a 'just get it working' patch; it's not compatible with OpenCV 3.x in its current state. Do we want to keep that (and OpenCV 2.x support, for that matter)? Or would we prefer to just make a hard switch?

**Comment by mleotta** [comment]
My suggestion, unless anyone objects, would be to remove OpenCV 2 support in KWIVER and instead support both 3 and 4 for now (similar to how we support both 2 and 3 currently).  The last person I can remember claiming he still needed OpenCV 2 was @mattdawkins, but that was long ago.  So if @mattdawkins no longer needs it for his work then I think we can drop OpenCV 2.

(if he does still need OpenCV 2 then we need to start to start twisting his arm harder to fix that 😉)

**Comment by mleotta** [comment]
Also, if we are supporting both OpenCV 3 and 4 then we should make 4 the default.  This will get people using it to see if there are issues, but they can easily revert to OpenCV 3 if there are issue that are not easy to fix.

**Comment by dstoup** [comment]
> Also, if we are supporting both OpenCV 3 and 4 then we should make 4 the default. This will get people using it to see if there are issues, but they can easily revert to OpenCV 3 if there are issue that are not easy to fix.

The problem with making 4 the default for now, even if kwiver builds is, some of the other packages in fletch that rely on OpenCV don't build with v4. It's possible that we can update some of those packages to grab any recently release OCV 4 support they added. But, that can take some time to sort through.

I am ok making it default on this PR and waiting until we get everything fixed or we can merge now and start pushing through the builds to get them working and update the default when that's done .. I will leave that up to you all since I don't know if there are pressing needs for OCV4 or not. Either way, approved pending your decision!

**Comment by mleotta** [comment]
I don't think there is a pressing need for OpenCV 4.  We should get it added as an option and make sure KWIVER can build against it, but we can leave OpenCV 3 as the default if there are conflicts.  What other packages in Fletch require OpenCV and do not support OpenCV 4?

**Review by dstoup — APPROVED** [review]

**Comment by dstoup** [comment]
Typical Windows dashboard failure which I guess I should look into .. merging

**Pull request merged**
